### PR TITLE
Enumeration support

### PR DESCRIPTION
**Pull request by htmldoug**
Implements the `scala.Enumeration` piece of #12.

Borrows heavily from https://github.com/plokhotnyuk/jsoniter-scala/pull/499.

@russellremple @perryjp @vito-c

### CODE REVIEW

**Review by jducoeur — APPROVED** [review]
Looks good.  I assume that the pickler works even if it's defined outside of the Enumeration?  (I don't see any reason why it wouldn't, but I would expect it to be a fairly common use case, so...)

**Comment by htmldoug** [comment]
Added a test for it.

File: weepickle/src/com/rallyhealth/weepickle/v1/EnumFromTos.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.rallyhealth.weepickle.v1
+
+import java.util.concurrent.ConcurrentHashMap
+
+import com.rallyhealth.weepickle.v1.core.Types
+import com.rallyhealth.weepickle.v1.implicits.{DefaultFroms, DefaultTos}
+
+trait EnumFromTos extends Types with DefaultTos with DefaultFroms {
```
Comment (htmldoug) [inline weepickle/src/com/rallyhealth/weepickle/v1/EnumFromTos.scala]: Downside of a separate trait is that I can't add any new methods to it without breaking binary compat. One new trait per PR isn't sustainable.

I'm considering just dumping these into the abstract class.

**Comment by perryjp** [comment]
Is there a way to use this for default values when an enum isn't found? eg, if enums are "a", "b", "c", and we want "lkajsdf" to be "c"?

**Comment by htmldoug** [comment]
Not at the moment.

Would you also want `null`, `{"pony": true}`, and `[]` to default to "c" as well? If so, maybe we could generalize this into a `.recover(...)`, although that would probably require buffering to generalize. Probably not a major issue though if used only in tight scopes, like expected enum values.

**Comment by perryjp** [comment]
I think there could be an argument for `null` but I would say complex types (array, object) would fail. I think there's a difference between "I didn't find your value, here's a default" and "you didn't even try to give me something that could be an enumeration"

**Comment by perryjp** [comment]
The idea of a `recover` is interesting, what's the use case for buffering? I agree that in tight scopes it's probably not a major issue but once available, people will invariably abuse it without understanding the implications.

**Review by perryjp — APPROVED** [review]

**Review by perryjp — APPROVED** [review]
Might be interested in figuring out how to default to an enumeration value but this is still good to move forward with

**Pull request merged**